templates/verifier_groth16.vy.ejs

```diff
@@ -0,0 +1,189 @@
+# @version >=0.3.4
+
+# The prime q in the base field F_q for G1 
+BASE_MODULUS : constant(uint256) = 21888242871839275222246405745257275088696311157297823662689037894645226208583
+
+# The prime moludus of the scalar field of G1.
+SCALAR_MODULUS : constant(uint256) = 21888242871839275222246405745257275088548364400416034343698204186575808495617
+
+# the number of inputs
+N_INPUTS: constant(uint256) = <%=IC.length-1%> 
+
+struct G1Point:
+    X: uint256
+    Y: uint256
+
+# Encoding of field elements is: X[0] * z + X[1]
+struct G2Point:
+    X: uint256[2]
+    Y: uint256[2]
+
+struct VerifyingKey:
+    alfa1: G1Point
+    beta2: G2Point
+    gamma2: G2Point
+    delta2: G2Point
+    ic: DynArray[G1Point, <%=IC.length%>]
+
+struct Proof:
+    A: G1Point
+    B: G2Point
+    C: G1Point
+
+
+# @dev return the generator of G1
+@internal
+@pure
+def P1() -> G1Point:
+    return G1Point({X: 1, Y: 2})
+
+
+# @dev return the generator of G2
+@internal
+@pure
+def P2() -> G2Point:
+    return G2Point({
+        X: [
+          11559732032986387107991004021392285783925812861821192530917403151452391805634,
+          10857046999023057135944570762232829481370756359578518086990519993285655852781
+        ],
+        Y: [
+          4082367875863433681332203403145435568316851327593401208105741076214120093531,
+          8495653923123431417604973247489272438418190587263600148770280649306958101930
+        ]}
+      )
+
+
+# @dev return r the negation of p, i.e. p.addition(p.negate()) should be zero.
+@internal
+@view
+def negate(p: G1Point) -> G1Point:
+    if (p.X == 0) and (p.Y == 0):
+        return G1Point({X: 0, Y: 0})
+
+    # Validate input or revert
+    assert (p.X < BASE_MODULUS) or (p.Y < BASE_MODULUS), "invalid proof"
+
+    # We know p.Y > 0 and p.Y < BASE_MODULUS.
+    return G1Point({X: p.X, Y: BASE_MODULUS - p.Y})
+
+
+# @dev return r the sum of two points of G1@return r the sum of two points of G1
+@internal
+@view
+def addition(p1: G1Point, p2: G1Point) -> G1Point:
+    # By EIP-196 all input is validated to be less than the BASE_MODULUS and form points
+    # on the curve.
+    a: uint256[2] = [p1.X, p1.Y]
+    b: uint256[2] = [p2.X, p2.Y]
+    result : uint256[2] = ecadd(a, b)
+    r : G1Point = G1Point({X: result[0], Y: result[1]})
+    return r
+
+
+# @dev return r the product of a point on G1 and a scalar, i.e.
+#      p == p.scalar_mul(1) and p.addition(p) == p.scalar_mul(2) for all points p.
+@internal
+@view
+def scalar_mul(p: G1Point, s: uint256) -> G1Point:
+    a: uint256[2] = [p.X, p.Y]
+    result: uint256[2] = ecmul(a, s)
+    r: G1Point = G1Point({X: result[0], Y: result[1]})
+    return r
+
+
+# @dev Asserts the pairing check
+#      e(p1[0], p2[0]) *  .... * e(p1[n], p2[n]) == 1
+#      For example pairing([P1(), P1().negate()], [P2(), P2()]) should succeed
+#      this function supports n=N_INPUTS maximum
+@internal
+@view
+def pairing_check_4(p1 : DynArray[G1Point, 4], p2: DynArray[G2Point, 4]) -> bool:
```

Comment:
this should take a static array since the function body assumes the length is 4. this would also match the solidity implementation of pairingProd4.